### ISSUE TITLE
opt: extend Min/Max -> Order+Limit rule to work on non-scalar GROUP BY (with no columns)

### ISSUE THREAD

**Issue by knz** [cockroachdb/cockroach#23149]
Sub-problem extracted from #23147.

Schema:
```sql
 create table d(
   user_id int not null, 
   doc_id int not null, 
   revision int not null,
   payload bytes, 
   primary key(user_id, doc_id, revision)); 
```

Plan:
```
t> explain(verbose) select user_id, doc_id, MAX(revision) as max_revision from d docs where user_id = 123 and doc_id = 456 group by user_id, doc_id;
+----------------+-------+--------+-------------+--------------------+-----------------------------------------------+------------------------------------------------------------+
|      Tree      | Level |  Type  |    Field    |    Description     |                    Columns                    |                          Ordering                          |
+----------------+-------+--------+-------------+--------------------+-----------------------------------------------+------------------------------------------------------------+
| group          |     0 | group  |             |                    | (user_id, doc_id, max_revision)               |                                                            |
|  │             |     0 |        | aggregate 0 | docs.user_id       |                                               |                                                            |
|  │             |     0 |        | aggregate 1 | docs.doc_id        |                                               |                                                            |
|  │             |     0 |        | aggregate 2 | max(docs.revision) |                                               |                                                            |
|  │             |     0 |        | group by    | @1-@2              |                                               |                                                            |
|  └── render    |     1 | render |             |                    | (user_id, doc_id, revision)                   | user_id=CONST; doc_id=CONST; revision!=NULL; key(revision) |
|       │        |     1 |        | render 0    | docs.user_id       |                                               |                                                            |
|       │        |     1 |        | render 1    | docs.doc_id        |                                               |                                                            |
|       │        |     1 |        | render 2    | docs.revision      |                                               |                                                            |
|       └── scan |     2 | scan   |             |                    | (user_id, doc_id, revision, payload[omitted]) | user_id=CONST; doc_id=CONST; revision!=NULL; key(revision) |
|                |     2 |        | table       | d@primary          |                                               |                                                            |
|                |     2 |        | spans       | /123/456-/123/457  |                                               |                                                            |
+----------------+-------+--------+-------------+--------------------+-----------------------------------------------+------------------------------------------------------------+
```

I would have expected that the aggregation be simplified/eliminated to a single revscan with limit 1 (the max is the last row).

Even if we don't like revscans, if I change the schema to have `revision DESC` in the pk definition (incidentally, this is what #23147 does), I still don't get a forward scan with limit 1.

@RaduBerinde what do you think?

**Comment by RaduBerinde** [comment]
In general, we can't apply this optimization when we have GROUP BY. In this particular case, the GROUP BY clause does nothing. I'd say the right thing to do would be to remove any group by columns known to be constant, and then the optimization should kick in.

**Comment by knz** [comment]
I have renamed the issue according to your suggestion. I suppose it is still a worthwhile optimization, even though there is a workaround in the special case in #23147).

**Comment by andy-kimball** [comment]
The cost-based optimizer does remove functionally dependent `GROUP BY` columns. However, the `Max` => `Limit` rule currently only applies to `ScalarGroupBy`. It could be extended to work for a `GroupBy` with no grouping columns as well.

**Comment by knz** [comment]
fixed in 21.1
```
demo@127.0.0.1:26257/defaultdb> explain(verbose) select user_id, doc_id, MAX(revision) as max_revision from d docs where user_id = 123 and doc_id = 456 group by user_id, doc_id;
                      info
------------------------------------------------
  distribution: local
  vectorized: true

  • render
  │ columns: (user_id, doc_id, max_revision)
  │ estimated row count: 1 (missing stats)
  │ render max: revision
  │ render user_id: user_id
  │ render doc_id: doc_id
  │
  └── • revscan
        columns: (user_id, doc_id, revision)
        estimated row count: 1 (missing stats)
        table: d@primary
        spans: /123/456-/123/457
        limit: 1

```
